collector/timex.go

```diff
@@ -0,0 +1,193 @@
+			"Value of the status array bits.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		constant: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "loop_time_constant"),
+			"Phase-locked loop time constant.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		tick: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "tick_seconds"),
+			"Seconds between clock ticks.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		ppsfreq: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "pps_frequency"),
+			"Pulse per second frequency.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		jitter: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "pps_jitter_seconds"),
+			"Pulse per second jitter.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		shift: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "pps_shift_seconds"),
+			"Pulse per second interval duration.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		stabil: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "pps_stability"),
+			"Pulse per second stability.",
+			nil, nil,
+		), prometheus.CounterValue},
+		jitcnt: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "pps_jitter_count"),
+			"Pulse per second count of jitter limit exceeded events.",
+			nil, nil,
+		), prometheus.CounterValue},
+		calcnt: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "pps_calibration_count"),
+			"Pulse per second count of calibration intervals.",
+			nil, nil,
+		), prometheus.CounterValue},
+		errcnt: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "pps_error_count"),
+			"Pulse per second count of calibration errors.",
+			nil, nil,
+		), prometheus.CounterValue},
+		stbcnt: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "pps_stability_exceeded_count"),
+			"Pulse per second count of stability limit exceeded events.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		tai: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "tai_offset"),
+			"International Atomic Time (TAI) offset.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		syncStatus: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "sync_status"),
+			"Is clock synchronized to a reliable server (1 = yes, 0 = no).",
+			nil, nil,
+		), prometheus.GaugeValue},
+	}, nil
+}
+
+func (c *timexCollector) Update(ch chan<- prometheus.Metric) error {
+	var syncStatus float64
+	var divisor float64
+	var timex = new(syscall.Timex)
+
+	status, err := syscall.Adjtimex(timex)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve adjtimex stats: %v", err)
+	}
+
+	if status == timeError {
+		syncStatus = 0
+	} else {
+		syncStatus = 1
+	}
+	if (timex.Status & staNano) != 0 {
+		divisor = nanoSecons
```

Comment:
Would be nicer to use the time package for doing these conversions. Could use time.Second here or in general change the time conversion to use the time package. No big issue though, if it's easier this way that's fine. But probably should use at least the time.Second constent here.

Reply:
Typo fixed, and comment improved. I tried to use time package, but to my C programmer eye it looked more confusing than clarifying. The trouble is that prometheus metric needs to be float64 and time units are int64 types. I don't know how to make them work without type casting allover, and that is what to me looks worse than a simple stupid constant.

Reply:
Okay fair enough